### PR TITLE
Pcp schedule sharing - frontend+backend

### PR DESCRIPTION
**Pull request by mialana**

### CODE REVIEW

File: frontend/plan/components/syncutils.js [hunk 1]
```diff
@@ -29,144 +31,108 @@ const allPushed = (scheduleState) => {
...
-            .then((response) => response.json())
-            .then((options) => {
-                handleSemester(options.SEMESTER);
-            })
-            .catch(() => {
-                store.dispatch(
-                    openModal("SEMESTER_FETCH_ERROR", {}, "An Error Occurred")
-                );
-            });
-    });
```
Comment (andyjiang3) [inline frontend/plan/components/syncutils.js:65]: why is this not needed anymore?

File: frontend/plan/components/syncutils.js [hunk 2]
```diff
@@ -29,144 +31,108 @@ const allPushed = (scheduleState) => {
...
-                                // find the name of the schedule with the deleted id
-                                const schedName = Object.entries(
-                                    scheduleState.schedules
+                    Object.keys(scheduleState.schedules).forEach(
+                        (scheduleName) => {
+                            if (
+                                !schedulesFromBackend.reduce(
+                                    (acc, schedule) =>
+                                        acc || schedule.name === scheduleName,
+                                    false
```
Comment (andyjiang3) [inline frontend/plan/components/syncutils.js:78]: `find` instead

File: frontend/plan/components/syncutils.js [hunk 3]
```diff
@@ -29,144 +31,108 @@ const allPushed = (scheduleState) => {
...
+                    );
+
+                    store.dispatch(
+                        checkForDefaultSchedules(schedulesFromBackend)
+                    );
+
+                    if (
+                        !scheduleState.primaryScheduleId ||
+                        scheduleState.primaryScheduleId === "-1" ||
+                        !schedulesFromBackend.reduce(
```
Comment (andyjiang3) [inline frontend/plan/components/syncutils.js:99]: `find`

File: frontend/plan/reducers/friendships.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
+import {
+    UPDATE_FRIENDSHIPS_ON_FRONTEND,
+    SWITCH_ACTIVE_FRIEND,
+} from "../actions/friendshipUtil";
+
+const initialState = {
+    pulledFromBackend: false,
+    activeFriend: {},
```
Comment (andyjiang3) [inline frontend/plan/reducers/friendships.js:8]: activeFriend refers to the friend's schedule the user is viewing? if so, consider renaming to something more clear

File: frontend/plan/types.ts [hunk 5]
```diff
@@ -166,6 +166,14 @@ export interface User {
     profile: Profile;
 }
 
+export interface Friendship {
+    sender: User;
+    recipient: User;
+    sent_at: string;
+    accepted_at: string;
+    status: string;
```
Comment (andyjiang3) [inline frontend/plan/types.ts:174]: maybe a ISO_date type for `sent_at` and `accepted_at`?

File: frontend/plan/components/schedule/ScheduleDisplay.tsx [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+const hashString = (s: string) => {
+    let hash = 0;
+    if (!s || s.length === 0) return hash;
+    for (let i = 0; i < s.length; i += 1) {
+        const chr = s.charCodeAt(i);
+        hash = (hash << 5) - hash + chr;
+        hash |= 0; // Convert to 32bit integer
+    }
+    return hash;
+};
```
Comment (andyjiang3) [inline frontend/plan/components/schedule/ScheduleDisplay.tsx:63]: move to utils

File: frontend/plan/components/schedule/ScheduleDisplay.tsx [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+
+    // 15 minute time intervals
+    const getNumRows = () => (endHour - startHour + 1) * 4 + rowOffset;
+    const getNumCol = () => 5 + colOffset + (showWeekend ? 2 : 0);
+
+    // step 2 in the CIS121 review: hashing with linear probing.
+    // hash every section to a color, but if that color is taken, try the next color in the
+    // colors array. Only start reusing colors when all the colors are used.
+    const getColor = (() => {
+        const colors = [
```
Comment (andyjiang3) [inline frontend/plan/components/schedule/ScheduleDisplay.tsx:185]: move to utils

File: frontend/plan/actions/index.js [hunk 8]
```diff
@@ -179,6 +166,32 @@ export const clearSchedule = () => ({
...
+    type: SET_PRIMARY_SCHEDULE_ID_ON_FRONTEND,
+});
+
+export const checkForDefaultSchedules = (schedulesFromBackend) => (
+    dispatch
+) => {
+    if (
+        !schedulesFromBackend.reduce(
+            (acc, { name }) => acc || name === "cart",
+            false
```
Comment (andyjiang3) [inline frontend/plan/actions/index.js:180]: `find`

File: frontend/plan/reducers/schedule.js [hunk 9]
```diff
@@ -1,119 +1,41 @@
...
-/**
- * Returns a copy of th given schedule without its id
- * @param schedule
- */
-const withoutId = (schedule) => {
-    const newSchedule = { ...schedule };
-    delete newSchedule.id;
-    return newSchedule;
+    readOnly: false,
+    primaryScheduleId: "-1",
```
Comment (andyjiang3) [inline frontend/plan/reducers/schedule.js:38]: is there a reason why `primaryScheduleId` is a string?

File: frontend/plan/reducers/schedule.js [hunk 10]
```diff
@@ -154,246 +76,148 @@ const nextAvailable = (scheduleName, used) => {
...
-    );
+const handleUpdateSchedulesOnFrontend = (state, schedulesFromBackend) => {
+    let newState = { ...state };
+
+    schedulesFromBackend.forEach((scheduleFromBackend) => {
+        if (
+            newState.deletedSchedules.reduce(
+                (acc, schedule) => acc || schedule === scheduleFromBackend.name,
+                false
+            )
```
Comment (andyjiang3) [inline frontend/plan/reducers/schedule.js:140]: `find`

File: frontend/plan/reducers/schedule.js [hunk 11]
```diff
@@ -154,246 +76,148 @@ const nextAvailable = (scheduleName, used) => {
...
+        } else {
+            const foundSchedule = newState.schedules[scheduleFromBackend.name];
+            // If changes to the schedule are still syncing, ignore the requested update
+            if (
+                !foundSchedule ||
+                (foundSchedule &&
+                    foundSchedule.pushedToBackend &&
+                    cloudUpdated >= foundSchedule.updated_at &&
+                    foundSchedule.updated_at - cloudUpdated >=
+                        MIN_TIME_DIFFERENCE)
```
Comment (andyjiang3) [inline frontend/plan/reducers/schedule.js:170]: `foundSchedule.updated_at - cloudUpdated` would always be negative and thus this if statement would always be `false`?

**Review by andyjiang3 — CHANGES_REQUESTED** [review]
Nice work! Thanks for also refactoring / removing locally storing schedule data - left some comments on refactoring and naming for code readability. Besides from that, we just need to:

1. Throughly test schedule sharing - approving requests doesn't seem to be working for me locally (giving me 409 error).
      - Viewing friends schedule: no courses added to their schedule, not sharing a schedule
      - Loading friends schedule
2. Test schedule / cart changes syncing with backend - ex. add courses to cart, log out, and make sure its up-to-date after logging in, edge cases: cart, semester, new acc (cart, "Schedule")??
3. @mialana can you add a few sentence to the PR describing the changes to backend-frontend syncing / locally storing some schedule data on the frontend for future reference?

File: backend/tests/plan/test_primary_schedules.py [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+        # remove u2 as a friend
+        friendshipu2 = Friendship.objects.get(sender=self.u1, recipient=u2)
+        friendshipu2.delete()
+
+        # only have u1 and u3 now
+        response = self.client.get(primary_schedule_url)
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(len(response.json()), 2)
+        # print("4", response.json())
+        # self.assertEqual(response.data[0]["id"], self.s.id)
```
[the remark this answers is not in the document]
Comment (anshnagwekar) [reply backend/tests/plan/test_primary_schedules.py:130]: I checked and wasn't able to remember, but I think it has to do with the schedules not loading properly in the testing environment (so all of the primary schedule tests were not set up correctly). It might be better if we just directly test out the endpoint since it's only 2 methods anyways.

File: frontend/plan/components/schedule/ScheduleSelectorDropdown.tsx [hunk 13]
```diff
@@ -28,57 +17,106 @@ const DropdownButtonContainer = styled.div<{ isActive: boolean }>`
...
+}
+
+const FriendButton = ({
+    friendName,
+    isActive,
+    display,
+    removeFriend,
+}: FriendButtonProps) => (
+    <ButtonContainer isActive={isActive} onClick={display}>
+        <ButtonLabelContainer width={75}>{friendName}</ButtonLabelContainer>
```
Comment (andyjiang3) [inline frontend/plan/components/schedule/ScheduleSelectorDropdown.tsx:71]: Can we add the friend icon before the name

**Comment by andyjiang3** [comment]
TODO: @anshnagwekar @andyjiang3 remove reverse_func in doc_settings and potentially other places

**Comment by mialana** [comment]
Primary updates:
- State changes are now done directly on backend and then updated accordingly on frontend.
- Redux logic added for friendship states (i.e. reducers and actions).
